multistage build test failures

## Version of Singularity:

singularity version v3.0.0-alpha.2-272-g5fb15299, current master.

### Failing tests

Failing when running `make test` on opensuse leap 15.0. Seems to be failing during sif-related multi-stage builds.

        --- FAIL: TestBuildMultiStage/SIFToSIF/SIF (0.02s)
            build_test.go:200: ^[[33mWARNING:^[[0m Authentication token file not found : Only pulls of public images will succeed
                panic: runtime error: index out of range

                goroutine 1 [running]:
                github.com/sylabs/singularity/src/pkg/build/types/parser.doHeader(0xc0003d2000, 0x8074, 0xc0004b6f78, 0x948f, 0xc0003d2000)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:290 +0x3fc
                github.com/sylabs/singularity/src/pkg/build/types/parser.canGetHeader(0x120b900, 0xc00000e010, 0xc00000e010, 0x0, 0x0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:340 +0x3a8
                github.com/sylabs/singularity/src/pkg/build/types/parser.IsValidDefinition(0x7ffdb125a8c8, 0x1f, 0x0, 0x0, 0x0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:428 +0xa4
                github.com/sylabs/singularity/src/pkg/build.makeDef(0x7ffdb125a8c8, 0x1f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:391 +0xb5
                github.com/sylabs/singularity/src/pkg/build.NewBuild(0x7ffdb125a8c8, 0x1f, 0x7ffdb125a8a8, 0x1f, 0x110efb6, 0x3, 0x1150000, 0xc000481070, 0x1, 0x1, ...)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:67 +0x72
                github.com/sylabs/singularity/src/cmd/singularity/cli.glob..func5(0x19baac0, 0xc0004de080, 0x2, 0x2)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build.go:125 +0x446
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x19baac0, 0xc0000300e0, 0x2, 0x2, 0x19baac0, 0xc0000300e0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19bdca0, 0x0, 0xfd9940, 0x1202dd0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x19bdca0, 0x4, 0x1142402)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
                github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:107 +0x5b
                main.main()
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20

            build_test.go:201: unexpected failure: exit status 2

### 

       --- FAIL: TestBuildMultiStage/WritableToSIF/SIF (0.02s)
            build_test.go:200: ^[[33mWARNING:^[[0m Authentication token file not found : Only pulls of public images will succeed
                panic: runtime error: index out of range

                goroutine 1 [running]:
                github.com/sylabs/singularity/src/pkg/build/types/parser.doHeader(0xc0003da000, 0x8074, 0xc0004b6f78, 0x94e3, 0xc0003da000)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:290 +0x3fc
                github.com/sylabs/singularity/src/pkg/build/types/parser.canGetHeader(0x120b900, 0xc00000e010, 0xc00000e010, 0x0, 0x0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:340 +0x3a8
                github.com/sylabs/singularity/src/pkg/build/types/parser.IsValidDefinition(0x7ffe343268c8, 0x1f, 0x0, 0x0, 0x0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:428 +0xa4
                github.com/sylabs/singularity/src/pkg/build.makeDef(0x7ffe343268c8, 0x1f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:391 +0xb5
                github.com/sylabs/singularity/src/pkg/build.NewBuild(0x7ffe343268c8, 0x1f, 0x7ffe343268a8, 0x1f, 0x110efb6, 0x3, 0x1150000, 0xc000485070, 0x1, 0x1, ...)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:67 +0x72
                github.com/sylabs/singularity/src/cmd/singularity/cli.glob..func5(0x19baac0, 0xc0004dc0e0, 0x2, 0x2)
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build.go:125 +0x446
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x19baac0, 0xc0000300e0, 0x2, 0x2, 0x19baac0, 0xc0000300e0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19bdca0, 0x0, 0xfd9940, 0x1202dd0)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
                github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x19bdca0, 0x4, 0x1142402)
                        /home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
                github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:107 +0x5b
                main.main()
                        /home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20

            build_test.go:201: unexpected failure: exit status 2

--- a/src/cmd/singularity/cli/build.go
+++ b/src/cmd/singularity/cli/build.go
@@ -17,6 +17,7 @@ import (
 	"github.com/sylabs/singularity/src/docs"
 	"github.com/sylabs/singularity/src/pkg/build"
 	"github.com/sylabs/singularity/src/pkg/sylog"
+	"github.com/sylabs/singularity/src/pkg/syplugin"
 )
 
 var (
@@ -83,7 +84,10 @@ var BuildCmd = &cobra.Command{
 	Short:   docs.BuildShort,
 	Long:    docs.BuildLong,
 	Example: docs.BuildExample,
-	PreRun:  sylabsToken,
+	PreRun: func(cmd *cobra.Command, args []string) {
+		sylabsToken(cmd, args)
+		syplugin.Init()
+	},
 	// TODO: Can we plz move this to another file to keep the CLI the CLI
 	Run: func(cmd *cobra.Command, args []string) {
 		buildFormat := "sif"

--- a/src/cmd/singularity/cli/singularity.go
+++ b/src/cmd/singularity/cli/singularity.go
@@ -17,7 +17,6 @@ import (
 	"github.com/sylabs/singularity/src/docs"
 	"github.com/sylabs/singularity/src/pkg/buildcfg"
 	"github.com/sylabs/singularity/src/pkg/sylog"
-	"github.com/sylabs/singularity/src/pkg/syplugin"
 	"github.com/sylabs/singularity/src/pkg/util/auth"
 )
 
@@ -154,7 +153,6 @@ func handleEnv(flag *pflag.Flag) {
 func persistentPreRun(cmd *cobra.Command, args []string) {
 	setSylogMessageLevel(cmd, args)
 	updateFlagsFromEnv(cmd)
-	syplugin.Init()
 }
 
 // sylabsToken process the authentication Token

--- a/src/pkg/build/build.go
+++ b/src/pkg/build/build.go
@@ -327,31 +327,15 @@ func (b *Build) runBuildEngine() error {
 		return fmt.Errorf("failed to marshal config.Common: %s", err)
 	}
 
-	// Set PIPE_EXEC_FD
-	pipefd, err := syexec.SetPipe(configData)
+	starterCmd, err := syexec.PipeCommand(starter, progname, env, configData)
 	if err != nil {
-		return fmt.Errorf("failed to set PIPE_EXEC_FD: %v", err)
+		return fmt.Errorf("failed to create cmd type: %v", err)
 	}
 
-	env = append(env, pipefd)
+	starterCmd.Stdout = os.Stdout
+	starterCmd.Stderr = os.Stderr
 
-	// Create os/exec.Command to run starter and return control once finished
-	starterCmd := &exec.Cmd{
-		Path:   starter,
-		Args:   progname,
-		Env:    env,
-		Stdout: os.Stdout,
-		Stderr: os.Stderr,
-	}
-
-	if err := starterCmd.Start(); err != nil {
-		return fmt.Errorf("failed to start starter proc: %v", err)
-	}
-	if err := starterCmd.Wait(); err != nil {
-		return fmt.Errorf("starter proc failed: %v", err)
-	}
-
-	return nil
+	return starterCmd.Run()
 }
 
 func getcp(def types.Definition, libraryURL, authToken string) (ConveyorPacker, error) {

--- a/src/pkg/build/types/parser/deffile.go
+++ b/src/pkg/build/types/parser/deffile.go
@@ -287,6 +287,10 @@ func doHeader(h string, d *types.Definition) (err error) {
 		trimLine := strings.Split(line, "#")[0]
 
 		linetoks := strings.SplitN(trimLine, ":", 2)
+		if len(linetoks) == 1 {
+			return fmt.Errorf("header key %s had no val", linetoks[0])
+		}
+
 		key, val := strings.ToLower(strings.TrimSpace(linetoks[0])), strings.TrimSpace(linetoks[1])
 		if _, ok := validHeaders[key]; !ok {
 			return fmt.Errorf("invalid header keyword found: %s", key)
@@ -423,6 +427,13 @@ func IsValidDefinition(source string) (valid bool, err error) {
 	if err != nil {
 		return false, err
 	}
+
+	if s, err := defFile.Stat(); err != nil {
+		return false, fmt.Errorf("unable to stat file: %v", err)
+	} else if s.IsDir() {
+		return false, nil
+	}
+
 	defer defFile.Close()
 
 	ok, _ := canGetHeader(defFile)

--- a/src/pkg/syplugin/load.go
+++ b/src/pkg/syplugin/load.go
@@ -65,7 +65,7 @@ func registerPlugin(pl interface{}) {
 			sylog.Debugf("Registering plugin as type %s", plType)
 
 			if err := regFn(pl); err != nil {
-				sylog.Fatalf("Unable to register plugin??")
+				sylog.Fatalf("Unable to register plugin: %s", err)
 			}
 			regWait.Done()
 		}(plType, regFn)

--- a/src/plugins/apps/apps.go
+++ b/src/plugins/apps/apps.go
@@ -31,6 +31,17 @@ const (
 	sectionRun     = "apprun"
 )
 
+var (
+	sections = map[string]bool{
+		sectionInstall: true,
+		sectionFiles:   true,
+		sectionEnv:     true,
+		sectionTest:    true,
+		sectionHelp:    true,
+		sectionRun:     true,
+	}
+)
+
 const (
 	globalEnv94Base = `## App Global Exports For: %[1]s
 	
@@ -158,6 +169,10 @@ func getAppAndSection(ident string) (appName string, sectionName string) {
 	identSplit := strings.Split(ident, " ")
 
 	if len(identSplit) < 2 {
+		return "", ""
+	}
+
+	if _, ok := sections[identSplit[0]]; !ok {
 		return "", ""
 	}
 

